src/main.rs

```diff
@@ -0,0 +1,272 @@
+            }
+        }
+
+        // Wait before checking again:
+        thread::sleep(poll_interval_duration);
+    }
+}
+
+fn check_mounts(mount_statuses: &mut HashMap<String, MountStatus>, logger: &syslog::Logger) {
+    let mount_points = get_mounts::get_mount_points();
+
+    // Remove any mount status entries which are no longer in the current list of mountpoints:
+    mount_statuses.retain(|ref k, _| {
+        mount_points.iter().position(|i| *i == **k).is_some()
+    });
+
+    for mount_point in mount_points {
+        // Check whether there's a pending test:
+        if let Some(mount_status) = mount_statuses.get_mut(&mount_point) {
+            if mount_status.check_process.is_some() {
+                let child = mount_status.check_process.as_mut().unwrap();
+
+                match child.try_wait() {
+                    Ok(Some(status)) => {
+                        logger
+                            .info(format!(
+                                "Slow check for mount {} exited with {} after {} seconds",
+                                mount_point,
+                                status,
+                                mount_status.last_checked.elapsed().as_secs()
+                            ))
+                            .unwrap_or_else(handle_syslog_error);
+                        ()
+                    }
+                    Ok(None) => {
+                        logger
+                            .warning(format!(
+                                "Slow check for mount {} has not exited after {} seconds",
+                                mount_point,
+                                mount_status.last_checked.elapsed().as_secs()
+                            ))
+                            .unwrap_or_else(handle_syslog_error);
+                        continue;
+                    }
+                    Err(e) => {
+                        logger.err(format!(
+                            "Status update for hung check on mount {} returned an error after {} seconds: {}",
+                            mount_point,
+                            mount_status.last_checked.elapsed().as_secs(),
+                            e
+                        )).unwrap_or_else(handle_syslog_error);
+                        ()
+                    }
+                }
+            }
+        }
+
+        let mount_status = check_mount(&mount_point);
+
+        if mount_status.alive {
+            logger
+                .debug(format!("Mount passed health-check: {}", mount_point))
+                .unwrap_or_else(handle_syslog_error);
+        } else {
+            let msg = format!("Mount failed health-check: {}", mount_point);
+            eprintln!("{}", msg);
+            logger
+                .err(msg)
+                .unwrap_or_else(handle_syslog_error);
+        }
+
+        mount_statuses.insert(mount_point.to_owned(), mount_status);
+    }
+}
+
+fn check_mount(mount_point: &str) -> MountStatus {
```

Comment:
I would expect this function to return a Result or replace each unwrap with `expect("custom error message")`

Reply:
That's a good point — I'd been thinking about it as returning a structure which had the status, child process, check time, etc. but if I don't care about the last check time for successful mounts a `Result` would make the contract more obvious.

Reply:
Things like error-chain give you the option to easily add error messages and adapt error types from other libraries. This way, you can move a bunch of `println`s out of this code and make it more concise by just using `?` instead of unwrap and handling the error as high up the call tree as possible, which often leads to less nested code :)